### ISSUE TITLE
Shaping of qNegativeIntegratedPosteriorVariance and optimize_acqf_function [Bug]

### ISSUE THREAD

**Issue by wjmaddox** [pytorch/botorch#573]
# 🐛 Bug

There's an indexing error that occurs at the end of an `optimize_acqf` call when optimizing `qNegativeIntegratedPosteriorVariance`. The error results because `qNegativeIntegratedPosteriorVariance` reduces over an incorrect dimension --- an input of size `batch x n x d` produces an output of size `batch x n_mc_points`. 

Edit: An error occurs in initialization if you don't use an objective.

## To reproduce

** Code snippet to reproduce **
```python
import torch

from botorch.optim import optimize_acqf
from botorch.models import SingleTaskGP
from botorch.acquisition.active_learning import qNegIntegratedPosteriorVariance
from botorch.acquisition.objective import ScalarizedObjective
from botorch.sampling.samplers import SobolQMCNormalSampler

train_X = torch.rand(10, 2)
Y = 1 - (train_X - 0.5).norm(dim=-1, keepdim=True)  # explicit output dimension
Y += 0.1 * torch.rand_like(Y)
train_Y = (Y - Y.mean()) / Y.std()

model = SingleTaskGP(train_X, train_Y)

qmc_sampler = SobolQMCNormalSampler(num_samples=32)
bounds = torch.stack([torch.zeros(2), torch.ones(2)])
qnipv = qNegIntegratedPosteriorVariance(
    model=model, 
    mc_points=train_X, 
    sampler=qmc_sampler,
    objective=ScalarizedObjective(weights=torch.ones(1))
)
```

** Stack trace/error message **
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-10-75fdb3cbe1eb> in <module>
      5     num_restarts=1,
      6     raw_samples=10,  # used for intialization heuristic
----> 7     options={"batch_limit": 5, "maxiter": 200},
      8 )

~/Documents/GitHub/botorch/botorch/optim/optimize.py in optimize_acqf(acq_function, bounds, q, num_restarts, raw_samples, options, inequality_constraints, equality_constraints, fixed_features, post_processing_func, batch_initial_conditions, return_best_only, sequential, **kwargs)
    190     if return_best_only:
    191         best = torch.argmax(batch_acq_values.view(-1), dim=0)
--> 192         batch_candidates = batch_candidates[best]
    193         batch_acq_values = batch_acq_values[best]
    194 

IndexError: index 1 is out of bounds for dimension 0 with size 1
```

## Expected Behavior

You should probably be able to optimize qNIPV through optimize_acqf.

## System information

**Please complete the following information:**
torch 1.5
botorch '0.3.2.dev19+gb2752ae1.d20201007'
gpytorch 1.2.0

**Comment by saitcakmak** [comment]
This appears to be due to an irregular return shape by `qNegIntegratedPosteriorVariance`.
I am not familiar with `qNegIntegratedPosteriorVariance`, but for an input `X`, it appears to return a `X.shape[:-2] x 10` tensor as output (`10` here being the number of `mc_points`). This is in contrast with the acquisition functions that I am familiar with, which would return a `X.shape[:-2]` tensor. The `optimize_acqf` function when used with `return_best_only=True` expect this behavior, which is why you get this error. 

> Edit: An error occurs in initialization if you don't use an objective.

If this is referring to the `RuntimeError: cannot sample n_sample > prob_dist.size(-1) samples without replacement` in `initialize_q_batch`, this is also due to the return shape of `qNegIntegratedPosteriorVariance`, and happens when you use `num_restarts > 10`. If you use `num_restarts <= 10`, this error is not raised, however, it leads to an extra batch dimension for the `initial_conditions`, which leads to further issues with `optimize_acqf`. 

As far as I can tell, this is not the intended behavior of `qNegIntegratedPosteriorVariance`. It returns a `X.shape[:-2]` output when the `objective=None` and `sampler=None`, and the `optimize_acqf` works without any issues. When the `objective` is specified, it is still intended to return a `X.shape[:-2]` tensor. In your code, you are passing a `sampler` with `num_samples=32`, which causes issues when `objective=None`. My understanding is you are only supposed to use `num_samples>1` when you are using an `objective` that requires so.

TLDR; If you use `objective=None` and `sampler=None` with `qNegIntegratedPosteriorVariance`, everything works as intended. Otherwise, your `objective` and `sampler` need to be compatible to avoid problems with the return shape.

**Comment by wjmaddox** [comment]
Thanks for the detailed comment! I managed to get `qNegIntegratedPosteriorVariance` to work ultimately by adding another explicit reduction over the mean:

```python
# where res is the qNegIntegratedPosteriorVariance output
if self.objective is None:
    return res.mean(dim=0)
else:
    return res.mean(dim=-1)
```

However, the current behavior is not the intended behavior as the posterior variance should be getting reshaped properly before being returned so that one can optimize it. 

>  If you use num_restarts <= 10, this error is not raised, however, it leads to an extra batch dimension for the initial_conditions, which leads to further issues with optimize_acqf.

is the error message that I was seeing. 

> My understanding is you are only supposed to use num_samples>1 when you are using an objective that requires so.

At least my understanding of the acquisition function is that `num_samples > 1` generates multiple fantasy samples and then batch fantasy constructs them (thereby creating yet another batch dimension to be reduced over). It seems like you'd really want to use `num_samples > 1` to reduce variance in optimization.

**Comment by saitcakmak** [comment]
Hi @wjmaddox, you're very welcome! I think you have a small misunderstanding there. If you check `res` before taking the mean, you will notice that the values (variance) is identical across the outer most batches (corresponding to fantasies). Since you are generating the fantasies for a fixed `X`, each fantasy will have a different posterior mean, however, the posterior variance will be identical. This is because the posterior variance does not depend on the observations, and only depends on the prior kernel and `train_X`. So, using `sampler=None` or equivalently `num_samples=1` will give you the same output, and is computationally efficient. 

> However, the current behavior is not the intended behavior as the posterior variance should be getting reshaped properly before being returned so that one can optimize it.

I believe this is because `qNegIntegratedPosteriorVariance` expects the `objective` to reshape the output of posterior variance. When `objective=None`, the code expects `# if single-output, shape is 1 x batch_shape x num_grid_points x 1`, whereas, when objective is not `None`, it expects `# if multi-output + obj, shape is num_grid_points x batch_shape x 1 x 1`. So, if your objective does not perform the appropriate reshaping, it ends up taking the `mean` over the wrong dimension, and fails to `squeeze` another one. This then produces the wrong output shape, which leads to the errors you get.

> At least my understanding of the acquisition function is that num_samples > 1 generates multiple fantasy samples and then batch fantasy constructs them (thereby creating yet another batch dimension to be reduced over). It seems like you'd really want to use num_samples > 1 to reduce variance in optimization.

This is correct whenever your acquisition function depends on the posterior mean. `qNegIntegratedPosteriorVariance` appears to be an exception to the rule.

**Comment by wjmaddox** [comment]
Ahh, I was confusing the inner expectation in the acquisition function as an expectation over the posterior variance, requiring re-sampling rather than the more straightforward mean.

**Comment by Balandat** [comment]
Thanks @saitcakmak for the detailed analysis of what's going wrong here.

>  This is because the posterior variance does not depend on the observations, and only depends on the prior kernel and train_X. So, using sampler=None or equivalently num_samples=1 will give you the same output, and is computationally efficient.

This is indeed true the crux here. We use the `fantasize` method as a shortcut to do get the updated posterior variance at the input locations (rather than re-implementing the posterior variance inference). 

Note that the objective here is a `ScalarizedObjective`, which is different from an `MCObjective` - rather than computing the objective function on samples from the model posterior at `X`, this objective class transforms the posterior itself (which works here b/c affine transforms of MVNs remain MVNs...). 

> Ahh, I was confusing the inner expectation in the acquisition function as an expectation over the posterior variance, requiring re-sampling rather than the more straightforward mean.

Indeed, the marginalization here is over the fantasy models. Since the posterior variance does not depend on the fantasized values), there is no point in using more than one sample. The fact that we're using "samples" at all here is just out of convenience so we can use the `fantasize` method. Now if your acquisition function also depends on the posterior mean, then yes, using a larger number of fantasy samples will help reduce variance. 

I'll close this for now, but I'm open to ideas that could help make this less confusing.

**Comment by saitcakmak** [comment]
@Balandat, what would you think of checking the return shape and raising an error if it is not equal to `X.shape[:-2]`? As far as I understand, after the final `mean` and `squeeze` operations, it should be returning a batch shape output. The error message could tell the user to check for a possible mismatch between the `sampler` and the `objective`.

**Comment by Balandat** [comment]
that makes a lot of sense to me. We should be able to just add this to the `t_batch_mode_transform` decorator to cover most other acquisition functions as well.
